Add ports to enable for tendrl server and nodes

TimothyAsirJeyasing opened the pull request:
https://github.com/Tendrl/tendrl-ansible/issues/49
Signed-off-by: Timothy Asir J <tjeyasin@redhat.com>

TimothyAsirJeyasing commented:
@mbukatov , @r0h4n , @nthomas-redhat , @shtripat , Please review

shtripat commented on roles/tendrl-storage-node/tasks/firewalld.yml:
```diff
@@ -0,0 +1,19 @@
+---
+
+- name: Install firewalld
```
I understand this doesn't hamper the prots enabled for gluster bricks etc. It just adds extra one needed for  webhook?

mbukatov replied:
You are right, this would break glusterfs setup if firewalld is unconfigured and disabled.

mbukatov replied:
The issue in which I suggested how to approach this (https://github.com/Tendrl/tendrl-ansible/issues/49) states:

> when firewalld service is not up and running, stop the playbook(via assert) and ask for manual intervention

TimothyAsirJeyasing replied:
yes,In that case better we can skip firewall configuration for node and server if its already disabled. Although there is no such known issues/limitation in tendrl-server, there might be some app pkgs which are running to work in firewalld disabled mode could behave improperly. So as to avoid this, we can skip firewall configuration if its already in disabled in that machine.  
What is your view on this ?

cybernth reviewed (changes requested):
I reccomend to create XML file with firewall rules and updated directly into /etc/firewalld/. This will be much better as the future management will be easier

TimothyAsirJeyasing commented:
Ansible provides easy way of handling addition or deletion of services and ports either tcp or udp and also it will take care of adding it in the specific file. Where as if we use xml file we may need to add individual xml files for the services and has to take care for removal as well.

cybernth commented:
> Ansible provides easy way of handling addition or deletion of services and ports either tcp or udp and also it will take care of adding it in the specific file. Where as if we use xml file we may need to add individual xml files for the services and has to take care for removal as well.

make sense. Looks good

mbukatov commented:
I suggest to add firewalld xml files for at least all tendrl components we directly own and report feature request for all gluster related ones.

TimothyAsirJeyasing commented:
Sure, as per the discussion, i will send seperate patches for adding the service xml files for tendrl-http, tendrl-api and monitoring-integration. Which will be installed as a part of rpm installation. And i will remove any configuration for tendrl-http, tendrl-api and monitoring-integration from this patch.

TimothyAsirJeyasing commented:
@mbukatov , @nthomas-redhat I have updated the patch set, please review

cybernth commented on roles/tendrl-server/handlers/main.yml:
```diff
@@ -36,6 +36,11 @@
     name=tendrl-notifier
     state=restarted
 
+- name: reload firewalld
```
have you checked this https://github.com/Tendrl/monitoring-integration/pull/338#issuecomment-365751699 ?

TimothyAsirJeyasing replied:
i will remove reload from here!

cybernth commented on roles/tendrl-server/tasks/main.yml:
```diff
@@ -28,6 +28,7 @@
 - include: tendrl-ui.yml
 - include: tendrl-monitoring-integration.yml
 - include: tendrl-notifier.yml
+- include: firewalld.yml
```
This needs to moved up in the order. In a rare case where firewalld is not installed, it might create issues while installing tendrl-api etc as the /usr/lib/firewalld folder doesn't exists

TimothyAsirJeyasing replied:
i will add "required firewalld" in the tendrl-api and monitoring-integration patch.

cybernth replied:
I don't think that is required. Firewalld is expected to be available with base OS

cybernth commented on roles/tendrl-server/handlers/main.yml:
```diff
@@ -35,3 +35,8 @@
   service:
     name=tendrl-notifier
     state=restarted
+
```
Are we using this anywhere?

TimothyAsirJeyasing replied:
I am using this to restart firewalld after enabling other ports

nthomas-redhat replied:
ok

cybernth replied:
Use reload instead of restart

mbukatov replied:
I need to recheck this.

mbukatov replied:
I need to recheck this.

cybernth commented on roles/tendrl-storage-node/handlers/main.yml:
```diff
@@ -5,3 +5,8 @@
     name=tendrl-node-agent
     state=restarted
+
+- name: restart firewalld
```
Are we using this anywhere?

TimothyAsirJeyasing replied:
I am using this to restart firewalld after enabling port for grafana server
in roles/tendrl-server/tasks/firewalld.yml

cybernth replied:
Use reload instead of restart

TimothyAsirJeyasing replied:
Sure, however i have seen few times that reload does not works. But restart works fine also reload after restart works fine.

TimothyAsirJeyasing replied:
Reload is not working only restart is working fine. What is your opinion about keeping this as restart only for now and will enhance later.

mbukatov replied:
Hmm, `systemctl reload firewalld` does reload permanent rules and makes them active:

```
# firewall-cmd --add-service=http --permanent
# firewall-cmd --list-all --permanent # this shows the http service
# firewall-cmd --list-all # this doesn't show http service
# systemctl reload firewalld
# firewall-cmd --list-all # now http service is shown here
```

cybernth reviewed (changes requested):
verified?

mbukatov reviewed (changes requested):
Current solution would break already glusterfs cluster when the cluster is running without firewalld or with firewall managed with plain iptables.

mbukatov commented:
Also I see that firewalld service files (just merged today) are not referenced there.

TimothyAsirJeyasing commented:
yes,In that case better we can skip firewall configuration for node and server if its already disabled. Although there is no such known issues/limitation in tendrl-server, there might be some app pkgs which are running to work in firewalld disabled mode could behave improperly. So as to avoid this, we can skip firewall configuration if its already in disabled in that machine.
What is your view on this ?

mbukatov commented:
@TimothyAsirJeyasing could I rebase this PR against current master branch (to resolve a conflict) and squash few commits together while doing this?

Like this: https://github.com/mbukatov/tendrl-ansible/tree/firewalld_rebased

Note that the content is the same, I just rebased it on master branch and squashed the commits into just 2, as I really appreciate having commits with meaningful commit messages, keeping changes separate if it makes sense. Addressing the same change id multiple different commits makes it harder for me to check what is going on.

mbukatov commented:
@TimothyAsirJeyasing 

> yes,In that case better we can skip firewall configuration for node and server if its already disabled.

I would rather abort the playbook with explanation so that the admin can decide how to resolve the situation (see https://github.com/Tendrl/tendrl-ansible/issues/49 and https://github.com/Tendrl/tendrl-ansible/issues/49#issuecomment-367097340)

> Although there is no such known issues/limitation in tendrl-server, there might be some app pkgs which are running to work in firewalld disabled mode could behave improperly. So as to avoid this, we can skip firewall configuration if its already in disabled in that machine.

Yes, you are right. We should not distinguish between tendrl-server machine and storage nodes, and use the same approach to make sure we don't break anything.

TimothyAsirJeyasing committed:
remove workaround.disable-firewall.yml playbook

tendrl-bug-id: https://github.com/Tendrl/tendrl-ansible/issues/49

TimothyAsirJeyasing committed:
Add ports to enable for tendrl server and nodes

tendrl-bug-id: https://github.com/Tendrl/tendrl-ansible/issues/49

mbukatov commented:
I'm going to add commits with suggested checks here (I was investigating how to do similar checks when working on prechecks playbook some time ago, so I can reuse this).

mbukatov committed:
1st impl. of firewalld assert in tendrl-server role

tendrl-bug-id: https://github.com/Tendrl/tendrl-ansible/issues/49

mbukatov commented:
I have implemented a firewalld assert in commit 33c882b86b830388af24f10adc6bf7d3a0585af2.

When the assert fails, the playbook is aborted with error message like this:

```
TASK [tendrl-server : Assert fail if firewalld is not installed and running] ***
Friday 23 February 2018  17:38:14 +0100 (0:00:00.430)       0:00:05.894 ********
fatal: [mbukatov-usm1-server.example.com]: FAILED! => {
    "assertion": "systemctl_is_enabled_firewalld.rc == 0",
    "changed": false,
    "evaluated_to": false,
    "msg": [
        "Admin intervention needed: tendrl-ansible was about to start",
        "opening Tendrl ports via firewalld but firewalld is either not",
        "installed, enabled nor running (see the assertion error above for",
        "details).",
        "You can either:",
        "1) Install firewalld and configure it to have ports opened for",
        "   services you are already running on the machines where you are",
        "   installing Tendrl (eg. Gluster). Then rerun this playbook.",
        "2) Set ansible variable 'configure_firewalld_for_tendrl' to False,",
        "   so that tendrl-ansible would not touch any firewall configuration",
        "   when you rerun the playbook.",
        "   In this mode, you own firewall configuration and are responsible",
        "   to make sure that all necessary ports are opened (including",
        "   Gluster and Tendrl ports).",
        "   For list of ports Tendrl requires to be open, refer to the",
        "   documentation."
    ]
}
        to retry, use: --limit @/home/martin/.ansible/retry-files/site.retry

PLAY RECAP *********************************************************************
mbukatov-usm1-client.example.com : ok=1 changed=0    unreachable=0    failed=0
mbukatov-usm1-gl1.example.com : ok=1    changed=0    unreachable=0    failed=0
mbukatov-usm1-gl2.example.com : ok=1    changed=0    unreachable=0    failed=0
mbukatov-usm1-gl3.example.com : ok=1    changed=0    unreachable=0    failed=0
mbukatov-usm1-gl4.example.com : ok=1    changed=0    unreachable=0    failed=0
mbukatov-usm1-gl5.example.com : ok=1    changed=0    unreachable=0    failed=0
mbukatov-usm1-gl6.example.com : ok=1    changed=0    unreachable=0    failed=0
mbukatov-usm1-server.example.com : ok=5 changed=0    unreachable=0    failed=1
```

@nthomas-redhat  @TimothyAsir @julienlim @a2batic  @jjkabrown1 please check if you find the message clear

mbukatov committed:
check firewalld in both roles via include_role

Move firewalld checks into dedicated task file and use `include_role`
module (since ansible 2.2) to include it in both tendrl-server and
tendrl-storage-node roles, avoiding unnecessary duplication.

tendrl-bug-id: https://github.com/Tendrl/tendrl-ansible/issues/49

mbukatov committed:
configure_firewalld_for_tendrl variable

tendrl-bug-id: https://github.com/Tendrl/tendrl-ansible/issues/49

mbukatov commented on roles/tendrl-server/tasks/firewalld.yml:
```diff
@@ -0,0 +1,44 @@
+    permanent=yes
+    state=enabled
+  notify:
+    - restart firewalld
```
Strictly speaking, we should have this notification in every task which enables a port.

mbukatov commented on roles/tendrl-server/tasks/firewalld.yml:
```diff
@@ -0,0 +1,44 @@
+---
```
The other detail here to check is why we don't use firewalld services we included in some tendrl packages.

mbukatov reviewed (changes requested):
Few details remains to be rechecked.

julienlim commented:
@mbukatov 

The message overall has the right content, but probably could be less verbose and more direct.

E.g.

        "tendrl-ansible did not complete execution as firewalld is not configured."
        "installed, enabled nor running."
        "You can either:",
        "1) Install firewalld, and ensure ports are configured for existing services",
        "   (e.g. Gluster). Then rerun this playbook.",
        "2) Set ansible variable 'configure_firewalld_for_tendrl' to False,",
        "   so that tendrl-ansible does not modify existing firewall configurations",
        "   when you rerun the playbook.",
        "   With this setting, you are responsible for the firewall configuration",
        "   and ensure all necessary ports are opened (including Gluster and",
        "   Tendrl ports).",
        "   For the list of ports Tendrl requires to be opened, refer to the",
        "   documentation."

cybernth commented:
> TASK [tendrl-server : Assert fail if firewalld is not installed and running] ***
> Friday 23 February 2018  17:38:14 +0100 (0:00:00.430)       0:00:05.894 ********
> fatal: [mbukatov-usm1-server.example.com]: FAILED! => {
>     "assertion": "systemctl_is_enabled_firewalld.rc == 0",
>     "changed": false,
>     "evaluated_to": false,
>     "msg": [
>         "Admin intervention needed: tendrl-ansible was about to start",
>         "opening Tendrl ports via firewalld but firewalld is either not",
>         "installed, enabled nor running (see the assertion error above for",
>         "details).",
>         "You can either:",
>         "1) Install firewalld and configure it to have ports opened for",
>         "   services you are already running on the machines where you are",
>         "   installing Tendrl (eg. Gluster). Then rerun this playbook.",
>         "2) Set ansible variable 'configure_firewalld_for_tendrl' to False,",
>         "   so that tendrl-ansible would not touch any firewall configuration",
>         "   when you rerun the playbook.",
>         "   In this mode, you own firewall configuration and are responsible",
>         "   to make sure that all necessary ports are opened (including",
>         "   Gluster and Tendrl ports).",
>         "   For list of ports Tendrl requires to be open, refer to the",
>         "   documentation."
>     ]
> }
>         to retry, use: --limit @/home/martin/.ansible/retry-files/site.retry

@mbukatov , Looks goog to me.
We agreed to implement https://github.com/Tendrl/tendrl-ansible/issues/49#issuecomment-367097340 in our last discussion. When are we planning to do that?
@TimothyAsirJeyasing ^

mbukatov commented:
@nthomas-redhat all points from https://github.com/Tendrl/tendrl-ansible/issues/49#issuecomment-367097340 are already present in the pull request, with exception of reload and other details I'm pointing out in my last review https://github.com/Tendrl/tendrl-ansible/pull/70#pullrequestreview-99021524 - I'm going to recheck it today so that the pull request will be ready asap.

mbukatov committed:
polishing error message for firewalld check

Based on feedback from Ju Lim.

mbukatov commented:
@julienlim thank you for the feedback, I have updated the message.

mbukatov committed:
use firewalld reload instead of restart

mbukatov committed:
don't open port for https

This will be reintroduced when https setup is added, right now it
doesn't make sense to open this port when we don't have this automated
in tendrl-ansible.

cybernth reviewed (dismissed):
LGTM

mbukatov committed:
firewall: open port for tendrl-monitoring-integration

Based on *Tendrl firewall settings* wikipage:

https://github.com/Tendrl/documentation/wiki/Tendrl-firewall-settings

mbukatov committed:
use immediate=true instead of reload handler

mbukatov commented:
merged
